dbms/src/DataStreams/TiRemoteBlockInputStream.h

```diff
@@ -128,51 +132,54 @@ class TiRemoteBlockInputStream : public IProfilingBlockInputStream
-                throw Exception(result.resp->error().DebugString());
-            }
-            /// only the last response contains execution summaries
-            if (result.resp != nullptr)
-            {
-                if constexpr (is_streaming_reader)
+                if (result.meet_error)
                 {
-                    addRemoteExecutionSummaries(*result.resp, result.call_index, true);
+                    LOG_WARNING(log, "remote reader meets error: {}", result.error_msg);
+                    throw Exception(result.error_msg);
                 }
-                else
+                if (result.eof)
+                    return false;
+                if (result.resp != nullptr && result.resp->has_error())
                 {
-                    addRemoteExecutionSummaries(*result.resp, 0, false);
+                    LOG_WARNING(log, "remote reader meets error: {}", result.resp->error().DebugString());
+                    throw Exception(result.resp->error().DebugString());
                 }
-            }
+                /// only the last response contains execution summaries
+                if (result.resp != nullptr)
+                {
+                    if constexpr (is_streaming_reader)
+                    {
+                        addRemoteExecutionSummaries(*result.resp, result.call_index, true);
+                    }
+                    else
+                    {
+                        addRemoteExecutionSummaries(*result.resp, 0, false);
+                    }
+                }
+
+                const auto & decode_detail = result.decode_detail;
 
-            const auto & decode_detail = result.decode_detail;
-
-            size_t index = 0;
-            if constexpr (is_streaming_reader)
-                index = result.call_index;
-
-            ++connection_profile_infos[index].packets;
-            connection_profile_infos[index].bytes += decode_detail.packet_bytes;
-
-            total_rows += decode_detail.rows;
-            LOG_TRACE(
-                log,
-                "recv {} rows from remote for {}, total recv row num: {}",
-                decode_detail.rows,
-                result.req_info,
-                total_rows);
-            if (decode_detail.rows > 0)
-                return true;
-            // else continue
+                size_t index = 0;
+                if constexpr (is_streaming_reader)
+                    index = result.call_index;
+
+                ++connection_profile_infos[index].packets;
+                connection_profile_infos[index].bytes += decode_detail.packet_bytes;
+
+                total_rows += decode_detail.rows;
+                LOG_TRACE(
+                    log,
+                    "recv {} rows from remote for {}, total recv row num: {}",
+                    decode_detail.rows,
+                    result.req_info,
+                    total_rows);
+                if (decode_detail.rows > 0)
+                    return true;
```

Comment:
It seems that we need to check all results before returning true

Reply:
Good point, it works now because the nextResult implementation will ensure only the last result having rows > 0. I'll consider it with your merging two "while" suggestion together.